### PR TITLE
Implement new interface for materials and coefficients

### PR DESCRIPTION
**Pull request by cmacmackin**
This simplifies the creation of coefficients and their use by kernels. It also makes the interface for the end-user more like that in Moose.

Currently coefficients haven't been totally expunged from the code, as they are still used by BCs. This should also be changed so that Platypus is fully in-line with Moose for materials and BCs. However, BCs will be done in a separate pull request.

### CODE REVIEW

File: test/tests/kernels/curlcurl.i [hunk 1]
```diff
@@ -67,16 +67,19 @@
...
+  [Substance]
+    type = MFEMGenericConstantMaterial
+    prop_names = one
+    prop_values = 1.0
+  []
+  [Forcing]
+    type = MFEMGenericFunctionVectorMaterial
+    prop_names = VolumetricSourceCoefficient
+    prop_values = forcing_field
   []
```
Comment (alexanderianblair) [inline test/tests/kernels/curlcurl.i:87]: The forcing field here is not really a material property - effectively it is just a function put into a coefficient so MFEM can use as a a (volumetric) source field.

Is it worth creating an MFEMFunction to handle this instead, or otherwise allow functions to instatiate properties without a material? This is a similar nomenclature issue to the type we will have when passing functions to boundary conditions for (eg.) a Dirichlet BC setting temperature on a surface (which isn't a material, but will still need a FunctionCoefficient).

Comment (cmacmackin) [reply test/tests/kernels/curlcurl.i:87]: How is this handled within MOOSE itself? For boundary conditions, the MOOSE classes take a function name as a parameter and I was planning to refactor ours to do the same.

Comment (alexanderianblair) [reply test/tests/kernels/curlcurl.i:87]: Similar to boundary conditions, the kernels that use them take a function name (or names) - an example is [VectorBodyForce](https://github.com/idaholab/moose/blob/next/framework/src/kernels/VectorBodyForce.C), which is adding an equivalent integrator to Platypus's `MFEMVectorFEDomainLFKernel`.  

In fact - every time MOOSE uses a function that I can think of, MFEM would use a FunctionCoefficient (or VectorFunctionCoefficient).

Comment (cmacmackin) [reply test/tests/kernels/curlcurl.i]: OK, I've reverted those changes now.

File: include/materials/MFEMMaterial.h [hunk 2]
```diff
@@ -2,17 +2,22 @@
...
 public:
   static InputParameters validParams();
+  static std::vector<std::string> subdomainsToStrings(std::vector<SubdomainName> blocks);
+  static libMesh::Point pointFromMFEMVector(const mfem::Vector & vec);
 
   MFEMMaterial(const InputParameters & parameters);
   virtual ~MFEMMaterial();
 
-  virtual void storeCoefficients(platypus::Subdomain & subdomain) {}
+  const std::vector<SubdomainName> & getBlocks() const { return _block_ids; }
```
Comment (alexanderianblair) [inline include/materials/MFEMMaterial.h:18]: Could be worth splitting the methods (and the `_block_ids` property) out into its own class that we then inherit from in a future PR, like `BlockRestrictable` in MOOSE (include/interfaces/BlockRestrictable.h)?

We'll also want to get the list of blocks using similar methods for other objects that may be domain restricted (such as kernels), and this would save having to implement them multiple times. Boundary conditions will have a similar story with boundary attributes (similar to MOOSE's `BoundaryRestrictable`)

**Review by alexanderianblair — APPROVED** [review]
Nice update that makes creation of coefficients to represent material properties closer in user experience to native MOOSE (and much less error prone then Platypus's previous method. Wide set of unit tests included.

The Material methods to store and handle the arrays of blocks could be split out to be used by other objects that are block restrictable (such as kernels) - but this is not a blocker for this PR, which is suitable as-is.
